crates/shielder-scheduler-server/src/storage/dynamo_db.rs

```diff
@@ -0,0 +1,475 @@
+        }
+
+        Ok(requests)
+    }
+
+    async fn get_item_by_id(&self, id: u128) -> Result<Option<ScheduledRequest>, StorageError> {
+        let id_str = id.to_string();
+        let out = self
+            .client
+            .get_item()
+            .table_name(TABLE_NAME)
+            .key("id", AttributeValue::S(id_str))
+            .send()
+            .await
+            .map_err(|e| map_internal("get_item", e))?;
+
+        if let Some(item) = out.item() {
+            if let Some(AttributeValue::S(payload_attr)) = item.get("payload") {
+                let req: ScheduledRequest = serde_json::from_str(payload_attr).map_err(|e| {
+                    StorageError::Internal(format!(
+                        "Failed to deserialize scheduled request payload: {e}"
+                    ))
+                })?;
+                return Ok(Some(req));
+            }
+        }
+        Ok(None)
+    }
+
+    async fn put_full_request(&self, request: &ScheduledRequest) -> Result<(), StorageError> {
+        let payload = serde_json::to_string(request).map_err(|e| {
+            StorageError::Internal(format!("Failed to serialize request payload: {e}"))
+        })?;
+
+        let mut builder = self.client.put_item();
+        builder = builder
+            .table_name(TABLE_NAME)
+            .item("id", AttributeValue::S(request.id.to_string()))
+            .item(
+                "status",
+                AttributeValue::S(status_to_str(&request.status).to_string()),
+            )
+            .item(
+                "relay_after",
+                AttributeValue::N(request.relay_after.timestamp().to_string()),
+            )
+            .item(
+                "created_at",
+                AttributeValue::N(request.created_at.timestamp().to_string()),
+            )
+            .item(
+                "last_note_index",
+                AttributeValue::S(request.last_note_index.to_string()),
+            )
+            .item("payload", AttributeValue::S(payload));
+
+        if request.error_message.is_some() {
+            builder = builder.item(
+                "error_message",
+                AttributeValue::S(request.error_message.clone().unwrap()),
+            );
+        }
+        if let Some(processed_at) = request.processed_at {
+            builder = builder.item(
+                "processed_at",
+                AttributeValue::N(processed_at.timestamp().to_string()),
+            );
+        }
+
+        builder
+            .send()
+            .await
+            .map_err(|e| StorageError::Internal(format!("Failed to put item: {e}")))?;
+        Ok(())
+    }
+}
+
+impl StorageInterface for DynamoDb {
+    async fn insert_scheduled_request(
+        &self,
+        request: ScheduledRequest,
+    ) -> Result<(), StorageError> {
+        let payload = serde_json::to_string(&request)
+            .map_err(|e| StorageError::Internal(format!("Failed to serialize request: {e}")))?;
+
+        let mut builder = self
```

Comment:
Can we please extract preparing `builder` to a separate function, to reduce code duplication?